lorien/Misc/StringTemplating.gd

```diff
@@ -0,0 +1,205 @@
+			g_arg_list,
+			Parser.GrammarLiteral.new(")")
+		]
+	)
+
+	parse_grammar = Parser.Grammar.new()
+	parse_grammar.elements.append(g_func_call)
+	parse_grammar.elements.append(g_string)
+	
+# -------------------------------------------------------------------------------------------------
+func process_string(template: String) -> String:
+	var offset := 0
+	while true:
+		var found = _find_template_location(template.substr(offset))
+		if not found:
+			break
+		found = found as TemplateLocation
+
+		var substitution = "TEMPLATE_ERROR"
+		var parsed = _parse(found.template_text.strip_edges())
+		if parsed == null:
+			_log_error("Cannot parse template '%s'" % found.template_text)
+		else:
+			var s = _apply_filter(parsed)
+			if s is String:
+				substitution = s
+
+		template = template.substr(0, found.start) + substitution + template.substr(found.end)
+		offset = found.start + len(substitution)
+	return template
+
+# -------------------------------------------------------------------------------------------------
+func _find_template_location(s: String):
+	var start = s.find(TEMPLATE_START)
+	if start == -1:
+		return null
+	
+	var result := TemplateLocation.new()
+	result.start = start
+
+	var in_string = null
+	for i in range(start, len(s)):
+		if in_string is String:
+			if s[i] == in_string:
+				in_string = null
+		elif s[i] in ["'", '"']:
+			in_string = s[i]
+		else:
+			if s.substr(i).find(TEMPLATE_END) == 0:
+				result.end = i + len(TEMPLATE_END)
+				result.template_text = s.substr(
+					result.start + len(TEMPLATE_START), 
+					result.end - result.start - len(TEMPLATE_END) - + len(TEMPLATE_START)
+				)
+				return result
+	_log_error("String '%s' does contain unreadable template" % s)
+
+# -------------------------------------------------------------------------------------------------
+func _parse(s: String):
+	s = s.strip_edges()
+	var parsed = parse_grammar.parse(s)
+	if parsed is Parser.ParsedSymbol:
+		# Check that everything got consumed
+		if parsed.last_position != len(s):
+			return null
+	return parsed
+
+# -------------------------------------------------------------------------------------------------
+func _apply_filter(parsed: Parser.ParsedSymbol):
+	if parsed.name == "string":
+		return parsed.value
+	elif parsed.name == "func_call":
+		var func_name: String = parsed.subsymbols[0].value
+		if ! func_name in filters:
+			_log_error("Filter '%s' does not exist" % func_name)
+			return null
+		
+		parsed = parsed as Parser.ParsedSymbol
+		var parsed_args_list: Parser.ParsedSymbol = parsed.find_first_subsymbol("args")
+		var arg_values = []
+		for a in parsed_args_list.subsymbols:
+			if a.name in ["(", ")", ","]:
+				continue
+			arg_values.append(_apply_filter(a))
+		return filters[func_name].call_funcv(arg_values)
+	else:
+		return null
+
+# -------------------------------------------------------------------------------------------------
+func _selftest():
+	# Test template detection
```

Comment:
I read somewhere (I think in the roadmap?) that you were looking to integrate a unittest framework at some point. Since it is not there yet(?), I added a "selftest" which should do _some_ tests at least. It is called during `_init()` and does not produce any output when it succeeds.

Reply:
There is no unit testing yet. I know there are some Godot frameworks but i haven't really looked into it. That's fine...maybe add a FIXME/TODO comment saying this should be replaced by real unit tests when available.